### DISCUSSION TITLE
Memory allocation on nonlinearSystem/updateInitialGuessDB

### DISCUSSION THREAD

**Discussion by thiagoalberici** [OpenModelica/OpenModelica#9440] [Q&A]
Hi everyone!
I need help with a memory allocation problem.

I made a simulation with OMEdit v1.19.3 and I'm using [FMI4cpp](https://github.com/NTNU-IHB/FMI4cpp) to run it on my cpp program. 
My model uses the libraries:
* Modelica v3.2.3
* ModelicaServices v3.2.3
* Complex v3.2.3

The simulation works fine, but the program is allocating approximately 150 Mb/min of my ram memory.

Here's a valgrind memory leak log record (one of many) after 100 seconds of execution:
> at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
by 0x8A42D10: createValueElement (in /tmp/fmi4cpp_mpagro_taurus_98342319/binaries/linux64/mpagro_taurus.so)
by 0x8A2CAE0: updateInitialGuessDB (in /tmp/fmi4cpp_mpagro_taurus_98342319/binaries/linux64/mpagro_taurus.so)
by 0x8A2D3C4: solve_nonlinear_system (in /tmp/fmi4cpp_mpagro_taurus_98342319/binaries/linux64/mpagro_taurus.so)
by 0x8982B01: mpagro_taurus_eqFunction_869 (in /tmp/fmi4cpp_mpagro_taurus_98342319/binaries/linux64/mpagro_taurus.so)
by 0x89865EB: mpagro_taurus_functionODE (in /tmp/fmi4cpp_mpagro_taurus_98342319/binaries/linux64/mpagro_taurus.so)
by 0x8A198C9: internalGetEventIndicators (in /tmp/fmi4cpp_mpagro_taurus_98342319/binaries/linux64/mpagro_taurus.so)
by 0x8A1A02E: fmi2DoStep (in /tmp/fmi4cpp_mpagro_taurus_98342319/binaries/linux64/mpagro_taurus.so)
by 0x26B43F: fmi4cpp::fmi2::cs_library::step(void*, double, double, bool) (in /home/git/isobus-ecu/build/src/x64/source/isobus-ecu)
by 0x26BD6A: fmi4cpp::fmi2::cs_slave::step(double) (in /home/git/isobus-ecu/build/src/x64/source/isobus-ecu)
by 0x139B96: fmi::FMIRunner::Loop() (in /home/git/isobus-ecu/build/src/x64/source/isobus-ecu)
by 0x16B80D: rtos::Task::exec(void*) (in /home/git/isobus-ecu/build/src/x64/source/isobus-ecu)

LEAK SUMMARY:
> definitely lost: 9,013,232 bytes in 375,549 blocks
   indirectly lost: 6,008,620 bytes in 375,541 blocks
     possibly lost: 2,264 bytes in 20 blocks
   still reachable: 9,715,675 bytes in 4,178 blocks
                      of which reachable via heuristic:
                        stdstring          : 611 bytes in 13 blocks
        suppressed: 0 bytes in 0 blocks

**Comment by AnHeuermann** [comment]
You can suppress most (invalid) reports by ignoring everything from the used garbage collector, see https://github.com/OpenModelica/OpenModelica/wiki/Memcheck-with-Valgrind.

To get the lines of source-code you need to build with debug flags.

@phannebohm and myself fiddled around with the list structures some time ago. You could try again with the latest nightly build and see if that changes anything. If not please open an issue about this.

**Comment by thiagoalberici** [comment]
Thanks for the fast response!

I was able to reproduce the same behavior (although on a smaller scale) even with the Boehm GC suppressions.

Also, switching to the latest nightly version doesn't seem to affect it.

Unfortunately I can't share my fmu, but I was also able to reproduce the problem with the example in Fluid > Examples > Tanks > TanksWithOverflow.

I opened an issue as requested. 

But in the meantime, is there a way to force a memory free somehow? Should I be calling some method once in a while?
Maybe force a garbage collection?
My simulation have to run with a 100us step and the memory allocation is growing really fast.

**Reply by AnHeuermann** [reply]
> But in the meantime, is there a way to force a memory free somehow? Should I be calling some method once in a while?
Maybe force a garbage collection?

I don't think so. At least I don't know without investigating #9442 and than I could just solve it (I hope).
